codelists/actions.py

```diff
@@ -491,30 +492,116 @@ def export_to_builder(*, version, author):
     )
+    draft, _ = update_draft(draft=draft, source_version=version)
+    return draft
+
 
-    # Recreate each search.  This creates the SearchResults linked to the new draft.  We
-    # can't just copy them across, because the data may have been updated, and new
-    # matching concepts might have been imported.
-    #
-    # In future, we should be able to short-circuit this by keeping track of the release
-    # that version was created with.
-    for search in version.searches.all():
-        codes = do_search(version.coding_system, term=search.term, code=search.code)[
-            "all_codes"
-        ]
-        builder_actions.create_search(
+def update_draft(*, draft, source_version=None):
+    """
+    Update this draft version in case data have been updated, by:
+    1) recreating the searches, either using the searches from a source version
+       (for a new draft created by `export_to_builder`), or from the searches
+       already on the draft
+    2) checking for any new codes and ensuring the codeset is up to date
+
+    This is called on every GET for a draft, since we can't (currently) tell if the
+    version of a coding system used to create the draft initially is the current one.
+    However, coding systems don't get updated very frequently, so we cache that we've
+    checked it, so we don't have to redo it every time.
+    """
+    cached_updates = cache.get("updated_drafts", [])
+    updates = {}
+    if draft.pk not in cached_updates:
+        _recreate_draft_searches(source_version or draft, draft)
+        updates = _check_new_concepts(draft)
+        cached_updates.append(draft.pk)
+        # cache for 24 hrs
+        cache.set("updated_drafts", cached_updates, timeout=60 * 60 * 24)
+    return draft, updates
+
+
+def _recreate_draft_searches(source_version, draft):
+    """
+    Recreate each search.  This creates the SearchResults linked to the new/updated draft.  We
+    can't just copy them across, because the data may have been updated, and new
+    matching concepts might have been imported.
+    In future, we should be able to short-circuit this by keeping track of the release
+    that version was created with.
+    """
+    for search in source_version.searches.all():
+        codes = do_search(
+            source_version.coding_system, term=search.term, code=search.code
+        )["all_codes"]
+        builder_actions.create_or_update_search(
             draft=draft, term=search.term, code=search.code, codes=codes
         )
+    cache_hierarchy(version=draft)
 
-    # This assert will fire if new matching concepts have been imported.  At the moment,
-    # the builder frontend cannot deal with a CodeObj with status ?  if any of its
-    # ancestors are included or excluded.  We will have to deal with this soon but for
-    # now fail loudly.
-    assert not draft.code_objs.filter(status="?").exists()
 
-    cache_hierarchy(version=draft)
+def _check_new_concepts(draft):
```

Comment:
I'm surprised that a function called `_check_something` makes changes to the database.